modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/ChunkedFileReader.java

```diff
@@ -81,17 +81,21 @@ boolean hasNextChunk() {
      * file.
      *
      * @return Chunk data.
-     * @throws IllegalStateException If there are no more chunks to read.
      * @throws IOException If an I/O error occurs.
      */
     byte[] readNextChunk() throws IOException {
         if (!hasNextChunk()) {
-            throw new IllegalStateException("No more chunks to read");
+            throw new IOException("No more chunks to read");
         }
 
         int toRead = (int) Math.min(chunkSize, length - offset);
         byte[] data = new byte[toRead];
-        stream.read(data);
+        int read = stream.read(data);
+
+        if (read != toRead) {
+            throw new IOException("Failed to read chunk data from file: expected " + toRead + ", actual " + read + "]");
+        }
```

Comment:
This is not what I meant, you should read in a loop.
-1 would mean an unexpected end of file, but you don't check for it exclusively. Please use the API properly